### PR TITLE
Converted API unit tests from stdlib to gocheck

### PR DESCRIPTION
**Pull request by michelvocks**
**\- What I did**
Converted stdlib unit tests to gocheck for the API package.
Replaced all `testutil/assert` with gocheck asserts.

**\- How I did it**
Reused already existing examples by the integration tests which are using the gocheck suite.

**\- How to verify it**

**\- Description for the changelog**
Converted stdlib unit tests to gocheck for the API package.

fixes partially #26175 

**\- A picture of a cute animal (not mandatory but encouraged)**
hqdefault

Signed-off-by: Michel Vocks michelvocks@gmail.com

### CODE REVIEW

File: api/client/formatter/container_test.go [hunk 1]
```diff
@@ -85,57 +93,38 @@ func TestContainerPsContext(t *testing.T) {
...
+	for _, ca := range cases {
+		ctx = containerContext{c: ca.container, trunc: ca.trunc}
+		v := ca.call()
 		if strings.Contains(v, ",") {
-			compareMultipleValues(t, v, c.expValue)
-		} else if v != c.expValue {
-			t.Fatalf("Expected %s, was %s\n", c.expValue, v)
+			compareMultipleValues(c, v, ca.expValue)
+		} else if v != ca.expValue {
+			c.Fatalf("Expected %s, was %s\n", ca.expValue, v)
```
Comment (tonistiigi) [inline api/client/formatter/container_test.go]:
```
else {
  c.Assert(c, check.Equals, ca.expValue)
}
```

**Comment by tonistiigi** [comment]
Design LGTM
@vdemeester @cpuguy83 You ok with porting over all the unit tests?

**Comment by vdemeester** [comment]
@tonistiigi I share mostly the same point of view as @stevvooe. I don't see too much value in using `gocheck`, especially for unit tests. I'm :-1: on this :sweat_smile:

**Comment by cpuguy83** [comment]
Agree with @vdemeester. I don't think we're really gaining anything since we don't need setup/teardown, suites, or anything like that for unit tests (which go 1.7 adds some nice features into the stdlib for anyway).

What was the motivation for wanting to make this change?

**Comment by tonistiigi** [comment]
@vdemeester @cpuguy83 
Some pros:
- Consistency with integration tests
- We can get rid of our own `testutil` package
- Shorter, this PR is -120 lines and it doesn't even remove the unused helper code.
- Easier to read
- Better error messages

negs:
- Learning curve. But you need to understand it anyway for integration-cli.

I do agree that there is no need to make it a rule that every package needs to use `go-check`. But for some complicated packages that use a lot of helper code this is nicer. I think if you need to write your own `equals()` helper you should just use `go-check` instead.

**Comment by stevvooe** [comment]
Some pros:

> Consistency with integration tests

I'm not sure if this is a benefit. The suite-based unit tests aren't necessarily better -- someone just used a framework.

> We can get rid of our own testutil package

No matter what, we are going to have common custom matcher code.

> Shorter, this PR is -120 lines and it doesn't even remove the unused helper code.

Yes, but it doesn't test the same thing. Most of this is from changing `if err != nil` blocks into `c.Assert`. The same can be done with a help `checkErr(t, err, "some message")` function.

Please see the comment in https://github.com/docker/docker/issues/26175#issuecomment-245045650, but `c.Assert(err, IsNil)` is not the same thing as `err != nil`. Specifically, it treats the case where the interface value is not nil, but the value pointed to be the interface is as the same thing, but it is not what most people check.

> Easier to read

That is arguable. As a Go programmer, I find these tests much hard to read.

> Better error messages

This can be achieved with stretchr assertions or bespoke check functions. `stretchr/assert` does the same thing without having to drop out of the standard test toolchain.

When listing cons, please make sure to be honest about the actual cons.

One major problem when using the `integration-cli` tests is that one has to learn the gocheck way of matching tests, rather than just the regular tooling. Often times, I don't even bother running, as it is too time consuming to look this up each time I need to modify one of these tests.

I think we should take this debate to #26175.

**Comment by xiaods** [comment]
@michelvocks  please rebase code!

**Comment by michelvocks** [comment]
@xiaods I will continue my work as soon as a official decision has been made by the maintainers.

**Comment by tonistiigi** [comment]
Following the discussion in here and #26175 I'm going the close this as most maintainers are against it.

@michelvocks Sorry and thanks for helping out.

**Pull request closed without merging**